docs/txp.md

```diff
@@ -0,0 +1,458 @@
+
+###### Public key witness
+
+Witness `PkWitness` contains:
+
+* public key `pk`
+* signature `sig`
+
+We require:
+
+* Address is a `PubKey` address
+* Address spending data is created with public key `pk` (\*)
+* Signature `sig` is a valid signature of transaction (validity checked with `pk`)
+
+Note: signature inside witness is given for `TxSigData` (which is basically the same as `Hash Tx`)
+  with `SignTx` tag.
+
+Note (\*): even though address doesn't contain its spending data, it's easy
+to check whether given spending data corresponds to given
+address. Suppose we have an address `addr` and spending data
+`asd`. Let's say that `addr` has attributes `attrs` and type `t`. We
+can construct `Address'` from `attrs`, `t` and `asd` and compare its
+hash with the one from `addr`.
+
+###### Script witness
+
+Witness `ScriptWitness` contains:
+
+* validator script, `validator`
+* redeemer script, `redeemer`
+
+We require:
+
+* Address is a `Script` address
+* Address spending data is created with validator script same as `validator`
+* Both `redeemer` and `validator` scripts have the same version.
+* If `verifyAllIsKnown` is `True`, script version (which is same for
+    redeemer and validator) must be known.
+    - For CSL 1.0 it means that version must be equal to 0, it's the only version known to our software.
+* Plutus validation script built from redeemer and validator for
+    given tx must return `True`
+
+###### Redeem witness
+
+Witness `PkWitness` contains:
+
+* redeem public key `pk`
+* redeem signature `sig`
+
+We require:
+
+
+* Address is a `Redeem` address
+* Address spending data is created with redeem public key `pk`
+* Signature `sig` is a valid signature of transaction (validity checked with `pk`)
+
+
+###### Unknown witness
+
+* If `verifyAllIsKnown` is `True`, witness is invalid
+* If witness has unknown type with tag `t`, corresponding address must
+  have unknown type with tag `t`
+
+##### Sums check
+
+This check is omitted if (and only if):
+* There is an unknown transaction input
+
+For each input and output we know its value (i.e. associated balance). We compute sum of all inputs and all outputs.
+
+* If sum of outputs is greater than sum of inputs, tx is considered invalid
+* Otherwise, check succeeds, difference between sum of inputs and sum of outputs is considered to be *transaction fee*.
+
+##### Transaction fee check
+
+This check is omitted if (and only if) one of conditions met:
+* There is an unknown transaction input
+* All inputs correspond to `Redeem` addresses
+  - Implication of this rule is that one can redeem ADA without paying fee
+* Currently adopted tx fee policy is unknown
+
+For this check we know:
+
+* Transaction fee `txFee` (from *Sums check*)
+* Transaction size `txSize` (from *Transaction size check*)
+* Currently adopted tx fee policy (as part of `BlockVersionData`), which is of known type
+
+###### Size-linear policy
+Only policy supported by CSL v1.0 is `TxFeePolicyTxSizeLinear`.
```

Comment:
I'll change CSL to Cardano SL, because, well, this is document will be read by external reader, I don't think that using our internal abbreviations is a good idea.